nni/nas/evaluator/evaluator.py

```diff
@@ -3,58 +3,389 @@
+
+    #     trial_command_channel = get_default_trial_command_channel()
+    #     original_params = nni.trial._params
+    #     original_seq = nni.trial._intermediate_seq
+
+    #     try:
+    #         set_default_trial_command_channel(_EvaluatorMockTrialCommandChannel(model))
+    #         assert nni.get_next_parameter() is model
+    #         yield
+    #     finally:
+    #         set_default_trial_command_channel(trial_command_channel)
+    #         # NOTE: It might have some side effects on nni.trial._params.
+    #         #       Might cause the trial_available() to still return true after the context.
+    #         #       The following hack might address the problem.
+    #         nni.trial._params = original_params
+    #         nni.trial._intermediate_seq = original_seq
 
     @staticmethod
-    def _load(ir: Any) -> 'Evaluator':
-        evaluator_type = ir.get('type')
-        if isinstance(evaluator_type, str):
-            # for debug purposes only
-            for subclass in Evaluator.__subclasses__():
-                if subclass.__name__ == evaluator_type:
-                    evaluator_type = subclass
-                    break
-        assert issubclass(cast(type, evaluator_type), Evaluator)
-        return cast(Type[Evaluator], evaluator_type)._load(ir)
-
-    @abc.abstractmethod
+    def _load(**ir: Any) -> 'Evaluator':
+        """Subclass implements ``_load`` for their own serialization."""
+        raise NotImplementedError()
+
     def _dump(self) -> Any:
+        """Subclass implements ``_dump`` for their own serialization."""
+        raise NotImplementedError()
+
+    # def _execute(self, model: ExecutableModelSpace) -> Any:
+    def _execute(self, model: Any) -> Any:
+        """Advanced users can overwrite this to avoid instantiation of the deep learning model.
+
+        For internal uses only.
+        """
+        executable_model = model.executable_model()
+        return self.evaluate(executable_model)
+
+
+class MutableEvaluator(Mutable, Evaluator):
+    """
+    Evaluators with tunable parameters by itself (e.g., learning rate).
+
+    The tunable parameters must be an argument of the evaluator's instantiation,
+    or an argument of the arguments' instantiation and etc.
+
+    To use this feature, there are two requirements:
+
+    1. The evaluator must inherit :class:`MutableEvaluator` rather than :class:`Evaluator`.
+    2. Make sure the init arguments have been saved in ``trace_kwargs``,
+       and the instance can be cloned with ``trace_copy``.
+       The easiest way is to wrap the evaluator with :func:`nni.trace`.
+       If the mutable parameter exists somewhere in the nested instantiation.
+       All the levels must all be wrapped with :func:`nni.trace`.
+
+    Examples
+    --------
+    >>> def get_data(shuffle): ...
+    ...
+    >>> @nni.trace                                  # 1. must wrap here
+    ... class MyOwnEvaluator(MutableEvaluator):     # 2. must inherit MutableEvaluator
```

Comment:
under which case do not need `nni.trace` to use `MutableEvaluator`?

Reply:
Sometimes, if the evaluator is used for debugging only and not mutating, we might allow not using `nni.trace`. But in most cases, `nni.trace` **must** be used to decorate a `MutableEvaluator`. Otherwise users might get a series of errors complaining that the the object is not traceable.

The reason I didn't make `nni.trace` part of `MutableEvaluator` is:

1. It's not really possible technically.
2. I have to write some `TraceableMixin` to do that. I think it's not worth the effort because `MutableEvaluator` is rather advanced. In most cases, users use `FunctionalEvaluator` and `LightningEvaluator`, which are subclasses of `MutableEvaluator`, and they don't have to care about `nni.trace`.